src/platform/OpenThread/MdnsImpl.cpp

```diff
@@ -42,14 +44,24 @@ const char * GetProtocolString(MdnsServiceProtocol protocol)
 
 CHIP_ERROR ChipMdnsPublishService(const MdnsService * service)
 {
-    if (service == nullptr)
-        return CHIP_ERROR_INVALID_ARGUMENT;
+    CHIP_ERROR result = CHIP_NO_ERROR;
+
+    VerifyOrExit(service, result = CHIP_ERROR_INVALID_ARGUMENT);
 
     char serviceType[kMdnsTypeMaxSize + kMdnsProtocolTextMaxSize + 1];
     snprintf(serviceType, sizeof(serviceType), "%s.%s", service->mType, GetProtocolString(service->mProtocol));
 
-    return ThreadStackMgr().AddSrpService(service->mName, serviceType, service->mPort, service->mTextEntries,
-                                          service->mTextEntrySize);
+    // Try to remove service before adding it, as SRP doesn't allow to update existing services.
+    result = ThreadStackMgr().RemoveSrpService(service->mName, serviceType);
+
+    // Service should be successfully removed or not found (not exists).
+    VerifyOrExit((result == CHIP_NO_ERROR) || (result == Internal::MapOpenThreadError(OT_ERROR_NOT_FOUND)), );
+
+    result =
+        ThreadStackMgr().AddSrpService(service->mName, serviceType, service->mPort, service->mTextEntries, service->mTextEntrySize);
```

Comment:
Instead of remove and re-add, we should refactor `AddSrpService` to `AddOrUpdateSrpService`: update the service if the service with exactly the same service name and instance name is already published.

Reply:
Do you mean update by modifying existing `otSrpClientService` and call `otSrpClientAddService` for that instance once again? I tried to do that, but in that case `otSrpClientAddService` returns OT_ERROR_ALREADY and update fails.

Reply:
Also I tested it using `ot cli` and I spotted that there it is possible to update service, but by adding new `otSrpClientService` instance each time and calling `otSrpClientAddService` for new instance, not by modifying existing one. I don't think it is what I would like to do, as this way we may fill whole services list with following copies of the same instance, just because we want to update its content.

Reply:
> Do you mean update by modifying existing `otSrpClientService` and calling `otSrpClientAddService` for that instance once again? I tried to do that, but in that case `otSrpClientAddService` returns OT_ERROR_ALREADY and update fails.

Okay, it seems we should
1. have a `otSrpClientRefreshService(...)` to trigger re-sending of the updated service;
2. simply replace any added service entry with new one.

Reply:
Yes, that's correct. Don't get me wrong, but creating a list of services containing for example 5 different update versions of the same service with all TXT records content, name strings and etc. will waste lots of memory, so it will be unacceptable solution for some platforms, that can manage e.g. storing only one service at a time.